File: problems/0003_longest-substring-without-repeating-characters/Solution1.scala
object Solution {
    def lengthOfLongestSubstring(s: String): Int = {
        var (max, start, end) = (0, 0, 0)
        for (c <- s) {
            val x = find(s, c, start, end)
            end += 1
            x match {
                case Some(idx) =>
                    start = idx + 1
                case None =>
                    max = (end - start).max(max)
            }
        }
        max
    }
    @inline
    def find(s: String, c: Char, start: Int, end: Int): Option[Int] = {
        for (i <- start until end)
            if (s(i) == c)
                return Some(i)
        None
    }
}

File: problems/0003_longest-substring-without-repeating-characters/Solution1.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        var max = 0
        var start = 0
        var end = 0
        s.forEach { c ->
            val idx = find(s, c, start, end)
            ++end
            if (idx == -1)
                max = Math.max(max, end - start)
            else
                start = idx + 1
        }
        return max
    }
    inline fun find(s: String, c: Char, start: Int, end: Int): Int {
        for (i in start..(end - 1))
            if (s[i] == c)
                return i
        return -1
    }
}
